### PR TITLE
Fastq filter names

### PR DESCRIPTION
**Pull request by rimangan**
New option for fastqFilter. If you provide a line-delimited file containing read names, only fastq entries with names found in this file will be retained. For paired end reads, a read will be retained if either the forward or reverse read is found in the names file. Useful for pulling out all reads that map to a particular chromosome or construct, or those with other filter properties.

### CODE REVIEW

**Review by dasnellings — DISMISSED** [review]

**Review by RavenYL — APPROVED** [review]

**Review by craiglowe — COMMENTED** [review]
Looks really good!  I didn't see the code that implements the discardNames filter.  Maybe I just missed it?  I do think I saw your last two tests testing that feature, so it must be there.  Where does that filter get applied?

File: cmd/fastqFilter/fastqFilter.go [hunk 1]
```diff
@@ -90,8 +124,15 @@ func fastqFilter(s Settings) {
 			if len(i.Seq) > s.MaxSize {
 				continue
 			}
+			if s.RetainNamesList != "" {
```
[the remark this answers is not in the document]
Comment (rimangan) [reply cmd/fastqFilter/fastqFilter.go:135]: Oops! I was a commit behind, implemented now

**Review by craiglowe — COMMENTED** [review]
Great work!
I also see a bunch of bedFilter code, which I think has already been checked in?  Can we merge main into this branch?

File: cmd/fastqFilter/fastqFilter.go [hunk 2]
```diff
@@ -64,6 +92,21 @@ func fastqFilter(s Settings) {
...
+					if _, RevInMap = namesMap[i.Rev.Name]; !RevInMap {
+						continue //continue if neither the forward or reverse read is in the map
+					}
+				}
+			}
+			if s.DiscardNamesList != "" {
+				if _, FwdInMap = namesMap[i.Fwd.Name]; FwdInMap {
+					continue
+				}
+				if _, RevInMap = namesMap[i.Rev.Name]; !RevInMap {
```
Comment (craiglowe) [inline cmd/fastqFilter/fastqFilter.go:106]: I was expecting "RevInMap" instead of "!RevInMap" towards the end of the line.

**Review by craiglowe — COMMENTED** [review]
Wonderful to have this functionality.  I think I found one small bug

File: cmd/fastqFilter/fastqFilter.go (hunk 2 above)
Comment (rimangan) [reply cmd/fastqFilter/fastqFilter.go:106]: Great catch! Copy paste error. Fixed in next commit

**Review by craiglowe — APPROVED** [review]
Wonderful!

**Pull request merged**